### PR TITLE
Draft for value range analysis

### PR DESCRIPTION
**Pull request by serge-sans-paille**

### CODE REVIEW

File: content/2020-04-23-value-range-analysis.rst [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+:date: 2020-04-23
+:category: compilation
+:lang: en
+:authors: serge-sans-paille
+:summary: Pythran recently gained a significant improvment in its value range
+          analysis. Let's discuss some implementation details through a bunch of
+          examples.
+
+Not every storie begins with an issue, but this one does. And with a quite old
```
Comment (Chronum94) [inline content/2020-04-23-value-range-analysis.rst]: story

File: content/2020-04-23-value-range-analysis.rst [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+one! `#1059 <https://github.com/serge-sans-paille/pythran/issues/1059>`_ dates
+back to october, 2018 :-) At that time, I was trying to compile efficiently some
```
Comment (Chronum94) [inline content/2020-04-23-value-range-analysis.rst]: October, 'trying to efficiently compile'

File: content/2020-04-23-value-range-analysis.rst [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        r = _clip(r, 0, img.shape[0] - 1)
+        c = _clip(c, 0, img.shape[1] - 1)
+
+        r2 = _clip(r + rl, 0, img.shape[0] - 1)
+        c2 = _clip(c + cl, 0, img.shape[1] - 1)
+
+        ans = img[r, c] + img[r2, c2] - img[r, c2] - img[r2, c]
+        return max(0, ans)
+
+Very little change here: the type annotations disappear, as Pythran infers them
```
Comment (Chronum94) [inline content/2020-04-23-value-range-analysis.rst]: 'Very little changes here'

File: content/2020-04-23-value-range-analysis.rst [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            b += 1
+        if b == 9:
+            print("wtf")
+        if b == 10:
+            print("wtf")
+        if b == 11:
+            print("ok")
+        return a, b, c
+
+Running this code through ``pythran -P``, which basically optimizes the code
```
Comment (Chronum94) [inline content/2020-04-23-value-range-analysis.rst]: remove 'basically'

File: content/2020-04-23-value-range-analysis.rst [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+entered at least once, and Pythran knows that thanks to the ``assert``.
+
+And as Pythran doesn't perform any kind of modelisation of the accumulation in
+``a`` and ``b``, it cannot get smarter than this, but well, that's already a
+decent start ;-)
+
+Conclusion
+==========
+
+Use ``assert`` statement! Pythran can extract precious information from them,
```
Comment (Chronum94) [inline content/2020-04-23-value-range-analysis.rst]: 'statements'

**Review by Chronum94 — COMMENTED** [review]
Mostly grammatical changes for a very slightly more 'business casual' feel.

**Comment by serge-sans-paille** [comment]
cc @lsix, I added the content you were longing for ;-)

File: content/2020-04-23-value-range-analysis.rst [hunk 6]
```diff
@@ -126,6 +126,59 @@ Let's illustrate that analysis through an example:
...
+(it shrinks towards ``-inf``) and ``b`` (it grows toward ``+inf``).
+This maybe not super accurate, but it's a correct overestimate.
+So we decide that right before the test, we have:
+
+.. code::
+
+    -inf <= a <= inf
+    11 <= b <= inf
+
+It's safe to apply the condition at the entry point too, so let's constraint our
```
Comment (Chronum94) [inline content/2020-04-23-value-range-analysis.rst]: constrain

File: content/2020-04-23-value-range-analysis.rst [hunk 7]
```diff
@@ -126,6 +126,59 @@ Let's illustrate that analysis through an example:
...
+value, so we have:
+
+.. code::
+
+    1 <= a <= inf
+    10 <= b <= 10
+    10 <= c <= 10``
+
+Then comes a ``while`` statement. A while usually has two successors: its body,
+and the next statement. But in that case we evaluate the condition and see that
```
Comment (Chronum94) [inline content/2020-04-23-value-range-analysis.rst]: Change to `In that case...`

File: content/2020-04-23-value-range-analysis.rst (hunk 6 above)
Comment (Chronum94) [reply content/2020-04-23-value-range-analysis.rst]: Also, I'm not sure why it's safe to do this, so perhaps a tiny bit more detail here?

File: content/2020-04-23-value-range-analysis.rst [hunk 8]
```diff
@@ -143,13 +196,46 @@ then prints the python code back, gives:
...
+    def foo(a):
+        return 2
+
 
-And as Pythran doesn't perform any kind of modelisation of the accumulation in
-``a`` and ``b``, it cannot get smarter than this, but well, that's already a
-decent start ;-)
+Let's be honest, this algorithm is super greedy, and if we find a sequence of ``if``
+statements, it has an exponential complexity (and this happens a soon as we
+unroll a loop with a condition in its body). In that case we fall back to a
```
Comment (Chronum94) [inline content/2020-04-23-value-range-analysis.rst]: Is there a number-of-if's threshold where we switch? May be nice to mention here.

Comment (Chronum94) [reply content/2020-04-23-value-range-analysis.rst]: Also maybe link to the file where this analysis happens? Although I'm not too sure about that, so this is mostly open.

**Pull request merged**